### ISSUE TITLE
A newcomers experience with Haraka

### ISSUE THREAD

**Issue by hfossli-agens** [haraka/Haraka#2928]
Hi! 

First of all, thanks for developing Haraka. A really great tool. I am amazed by the community around it and the swift responses in the issues and pull requests. Great job. You are truly awesome. ❤️

I thought I could share my experience with Haraka. These are just my opinions and I don't expect everyone (or anyone?) to agree. Working with products daily I know how important it is to get feedback - that's why I'm sharing my thoughts here.

Even though it is easy to get the first hello world example up and running, there's quite a lot to learn because Haraka's setup is quite different from prior experiences with js frameworks. 

**Why isn't it more like express?** I wish all the .ini config files where replaced with a modular javascript framework much like express.

```js
const haraka = require('Haraka');

const server = haraka();

server.use(spamAssassin({ ... });

server.rcpt((next, conn, params) {
   const { rcpt } = params;
   console.log("Receiving from:", rcpt);
   next();
});

server.tls.listen(8025, {
   key: fs.readFileSync('../certs/my.domain.key'),
   cert: fs.readFileSync('../certs/my.domain.cer'),
}, function() {
   console.log("now listening on 8025");
});
```

**Why can't I control the entry point myself?** Booting up with `npx haraka -c .` feels alien and I loose control over everything. It would be great to be able to say `node server.js` which then loads and starts haraka.

**Why isn't it logging with console.log?** I think it would be better to control the redirection of logs outside the application than inside. I wish the default setup was with plain logging to console.log, but of course with the possibility to set log levels as that is useful no matter where the log is sent.

**Why can't I use pm2 or similar tools?** Haraka already does a lot with workers and pooling. I understand there are good reasons. I wish the default setup was simple without workers and pooling and then you could opt in. 

**Why can't I use ENV variables?** I tried giving [this](https://github.com/haraka/haraka-config/pull/50) a go, but Haraka didn't read read my tls.js file even though I had no tls.ini file. I really wish it was possible to inject configuration with ENV vars like [12-factor](https://12factor.net/config) strongly suggests for good reasons. 

--

I now know the answers to these initial questions I had and I understand there are good reasons for the way things are.

--

The bottom line for me is that Haraka does a lot of things *different*. Time is our scarcest commodity. An important rule of thumb when creating great user experiences is that we have to create something that feels natural to the user. What's natural is of course a subjective thing. Leveraging and playing upon users prior experiences is extremely important. That's why I think it is important that users can apply their knowledge from popular frameworks like express-js.

I think if Haraka was more like express, being a modular javascript framework, it would be much easier to use and I also think it would be great for growing the community around haraka.

This post can be closed without any further action.

**Comment by celesteking** [comment]
I think you've confused Haraka, which is an SMTP server, for nodemailer or similar. Although I see where you're clinging with all that programmatic-only configuration approach, I don't see any significant value in that for the "end-user" of an SMTP server, which is a sysadmin / devops. And those people like configuration files and don't like learning new coding languages only to write something that's easily laid out via, well, a text file.

As for the env vars, yeah, there's an open issue for that.

**Comment by hfossli-agens** [comment]
If config files is the goal you/haraka contributors have truly nailed it 😊 It is really easy. 

I once wrote a layout engine which read json files sent from server. It was great and the designers loved it. Still, what they always asked for was dynamic features only a programming language could offer. 

Anyways, thanks for hearing me out. Much love.
